### PR TITLE
Add custom logic to each `OutputResult` subclass [2/2]

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Follow-up to #9349

- Treat everything returned by the manual optimization `training_step` as an "extra".
- Customize `ManualResult` logic.
- Tighten output validation.
- Integrate output validation into the appropriate result class.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/batch/manual.py [hunk 1]
```diff
@@ -11,16 +11,50 @@
...
+
+    extra: Dict[str, Tensor] = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        # TODO: remove with the deprecation removal in v1.6
+        self._check_extra_detach_deprecation(self.extra)
+        self.extra = recursive_detach(self.extra)
+
+    @classmethod
+    def from_training_step_output(
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/manual.py:44]: This implementation treats everything returned as an "extra".

File: pytorch_lightning/loops/batch/manual.py [hunk 2]
```diff
@@ -11,16 +11,50 @@
...
-    check_finite_loss,
-)
+from pytorch_lightning.loops.closure import OutputResult
+from pytorch_lightning.loops.utilities import _build_training_step_kwargs, _extract_hiddens, check_finite_loss
+from pytorch_lightning.utilities.memory import recursive_detach
+from pytorch_lightning.utilities.types import STEP_OUTPUT
+
+
+@dataclass
+class ManualResult(OutputResult):
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/manual.py:27]: The `ManualResult` is defined alongside the `ManualLoop`

File: pytorch_lightning/loops/closure.py [hunk 3]
```diff
@@ -12,68 +12,22 @@
...
-        elif isinstance(training_step_output, Tensor):
-            closure_loss = training_step_output
-
-        if closure_loss is not None:
-            # accumulate the loss. If ``accumulate_grad_batches == 1``, no effect
-            closure_loss /= normalize
-
-        return cls(closure_loss, extra=extra)
-
+class OutputResult:
```
Comment (carmocca) [inline pytorch_lightning/loops/closure.py:26]: This is basically empty and just defined for the inheritance

File: pytorch_lightning/loops/closure.py [hunk 4]
```diff
@@ -12,68 +12,22 @@
...
     @staticmethod
     def _check_extra_detach_deprecation(extra: Dict[str, Any]) -> None:
+        # TODO: remove with the deprecation removal in v1.6
+        # this is only here to avoid duplication
```
Comment (carmocca) [inline pytorch_lightning/loops/closure.py:30]: just deduplicating code. not a big deal since it will be gone in v1.6

File: pytorch_lightning/loops/closure.py [hunk 5]
```diff
@@ -122,69 +71,9 @@ def consume_result(self) -> ClosureResult:
...
     @abstractmethod
-    def closure(self, *args: Any, **kwargs: Any) -> ClosureResult:
+    def closure(self, *args: Any, **kwargs: Any) -> OutputResult:
         """Implements the behavior of the closure once it is getting called."""
         pass
 
-    def __call__(self, *args: Any, **kwargs: Any) -> Optional[Tensor]:
+    def __call__(self, *args: Any, **kwargs: Any) -> None:
         self._result = self.closure(*args, **kwargs)
-        return self._result.loss
```
Comment (carmocca) [inline pytorch_lightning/loops/closure.py:131]: Notice that this no longer assumes that the `loss` attribute exists

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 6]
```diff
@@ -308,17 +308,10 @@ def _prepare_outputs(
...
+                if isinstance(batch_outputs, OutputResult):
                     batch_outputs = [batch_outputs]
 
-                for tbptt_output in batch_outputs:
-                    out = {}
-                    if tbptt_output.loss is not None:
-                        out["loss"] = tbptt_output.loss
-                    out.update(tbptt_output.extra)
-                    processed_tbptt_outputs.append(out)
+                processed_tbptt_outputs = [tbptt_output.extra for tbptt_output in batch_outputs]
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:314]: Broken, needs to be fixed.

We need different logic depending on the result class. Either that or a common interface to return what's processed.

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 7]
```diff
@@ -20,19 +21,143 @@
...
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 from pytorch_lightning.utilities.finite_checks import detect_nan_parameters
 from pytorch_lightning.utilities.imports import _TPU_AVAILABLE
+from pytorch_lightning.utilities.memory import recursive_detach
+from pytorch_lightning.utilities.types import STEP_OUTPUT
+from pytorch_lightning.utilities.warnings import WarningCache
+
+
+@dataclass
+class ClosureResult(OutputResult):
```
Comment (carmocca) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:43]: `ClosureResult` and `Closure` are now defined alongside the `OptimizerLoop` aka `AutomaticLoop`.

We could also create a folder for it

```python
optimization/
    manual/
        loop.py
        closure.py
    automatic/
        loop.py
        closure.py
```

File: pytorch_lightning/loops/utilities.py [hunk 8]
```diff
@@ -38,31 +38,6 @@ def check_finite_loss(loss: Optional[torch.Tensor]) -> None:
         raise ValueError(f"The loss returned in `training_step` is {loss}.")
 
 
-def _check_training_step_output(model: "pl.LightningModule", training_step_output: STEP_OUTPUT) -> None:
```
Comment (carmocca) [inline pytorch_lightning/loops/utilities.py:41]: Each result class now does this validation

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:43]: - I would define the specific result definitions directly in the loop files. 
- closure.py can't really go with manual_loop

Comment (carmocca) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:43]: > closure.py can't really go with manual_loop

oops yes, meant to write `result.py`

> I would define the specific result definitions directly in the loop files.

So keep it as it is?

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:314]: the training batch loop owns the optimizer loop and the manual loop. therefore it can be the training_batch_Loop's responsibility to retrieve the output and standardize it into a format that the higher-up loop expects. what do you think?

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:43]: yes, I think that would serve us the best

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 6 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:314]: I think we should split `_prepare_outputs` into pieces across loops. Right now the information of the subloops leaks into the epoch loop as this function tries to process everything at once.

If you have ideas around it, I'm all ears 😄

File: pytorch_lightning/loops/optimization/manual_loop.py [hunk 9]
```diff
@@ -144,15 +113,10 @@ def advance(self, batch: Any, batch_idx: int) -> None:  # type: ignore[override]
...
 
-            _check_training_step_output(lightning_module, training_step_output)
-
             self._hiddens = _extract_hiddens(training_step_output, lightning_module.truncated_bptt_steps)
 
             result = ManualResult.from_training_step_output(training_step_output, self.trainer.accumulate_grad_batches)
 
-            if self.trainer.terminate_on_nan:
-                check_finite_loss(result.closure_loss)
-
```
Comment (carmocca) [inline pytorch_lightning/loops/optimization/manual_loop.py:155]: Since the manual loop does not need to return a loss, this is no longer done. The user would need to call `check_finite_loss` manually.

This should be okay with the introduction of `detect_anomaly`

File: pytorch_lightning/loops/optimization/closure.py [hunk 10]
```diff
@@ -13,16 +13,37 @@
...
+                    f"One of the returned values {set(extra.keys())} has a `grad_fn`. We will detach it automatically"
+                    " but this behaviour will change in v1.6. Please detach it manually:"
+                    " `return {'loss': ..., 'something': something.detach()}`"
+                )
+                return v.detach()
+            return v
+
+        return apply_to_collection(extra, Tensor, check_fn)
+
+    def asdict(self) -> Dict[str, Any]:
```
Comment (tchaton) [inline pytorch_lightning/loops/optimization/closure.py:45]: Should this be abstract or raise a not ImplementedError. If someone forgets to override it, it could be complicated the trace the error.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 11]
```diff
@@ -312,18 +312,7 @@ def _prepare_outputs(
...
-                    batch_outputs = [batch_outputs]
-
-                for tbptt_output in batch_outputs:
-                    out = {}
-                    if tbptt_output.loss is not None:
-                        out["loss"] = tbptt_output.loss
-                    out.update(tbptt_output.extra)
-                    processed_tbptt_outputs.append(out)
-
+                processed_tbptt_outputs = batch_outputs if isinstance(batch_outputs, list) else [batch_outputs]
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:315]: great reduction here!

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
